### ISSUE TITLE
Device-Specific Tensor Dispatch  in Torchao

### ISSUE THREAD

**Issue by liangan1** [pytorch/ao#4801]
## Motivation

TorchAO is natively designed for CUDA. Adding an additional device in torchAO is challenge due to the device-specific logic. The design in this RFC is general and I will take the Intel XPU as the example to explain our design. 

Intel XPU is an in-tree PyTorch device. As described in the [XPU Enabling Plan](https://github.com/pytorch/ao/issues/3576), the goal is to provide a CUDA-like torchao user experience with nearly zero migration effort and to reuse torchao implementations wherever the CUDA and XPU behavior is equivalent.

The difficult part is handling cases where XPU behavior differs from CUDA behavior. The current enablement approach adds device checks to generic tensor implementations and workflows. A typical pattern is:

```python
if tensor.device.type == "xpu":
    # XPU layout or kernel path
else:
    # existing CUDA path
```

These branches are individually small, but they accumulate across tensor construction, layout conversion, operator dispatch, validation, serialization, and tests. This creates several problems:

1. Generic torchao files acquire backend-specific knowledge and become harder to reason about.
2. A change to one backend can regress another backend because both paths share the same functions and configuration state.
3. Device-specific changes require review and long-term maintenance from the owners of the generic implementation, even when the behavior is owned and tested by a hardware vendor.
4. Adding another backend repeats the same pattern and grows nested device conditionals.
5. Backend-specific files do not have a clear ownership boundary.

This RFC proposes an opt-in registration and dispatch mechanism for device-specific subclasses of `TorchAOBaseTensor`.

The proposal keeps torchao's public APIs, configuration objects, and generic tensor abstractions unchanged. A backend may register a device-specific subclass for an existing torchao tensor type. At tensor creation points, torchao looks up the implementation by the input device and constructs the registered subclass. If no implementation is registered, torchao constructs the existing tensor class exactly as it does today.

For Intel XPU, device-specific subclasses and operator overrides will live under an `xpu/` subdirectory and will be maintained by Intel. The generic implementation remains the source of shared semantics, while the XPU subclass contains only the differences required by XPU kernels, layouts, or runtime behavior.

## Proposed Design

This PR #4800 provides the changes of this general design which can be used by all AO Tensors. Besides, we take the MXTensor as a example to show the changes for the XPU in #4800. 

### Device-specific tensor registry

Add a registry next to `TorchAOBaseTensor`. The registry maps a generic torchao tensor class and a PyTorch device type to a device-specific subclass:

```python
_DEVICE_TENSOR_REGISTRY: dict[
    type[TorchAOBaseTensor],
    dict[str, type[TorchAOBaseTensor]],
] = {}


def register_ao_tensor(
    generic_type: type[TorchAOBaseTensor],
    device: str,
    specific_type: type[TorchAOBaseTensor],
) -> None:
    ...


def get_ao_tensor(
    generic_type: type[TorchAOBaseTensor],
    device: str,
) -> type[TorchAOBaseTensor]:
    return _DEVICE_TENSOR_REGISTRY.get(generic_type, {}).get(
        device, generic_type
    )
```

Registration must enforce the following invariants:

- `generic_type` is a `TorchAOBaseTensor` subclass.
- `specific_type` is a subclass of `generic_type`.
- `device` is normalized through `torch.device(device).type`.
- Duplicate registration for the same `(generic_type, device)` is rejected unless it registers the same class. This prevents import order from silently changing behavior.

Example registration:

```python
class MXTensorXPU(MXTensor):
    ...


register_ao_tensor(MXTensor, "xpu", MXTensorXPU)
```

### Dispatch at tensor creation boundaries

Code that creates an existing TorchAO tensor selects the implementation from the device of the source tensor:

```python
weight = getattr(module, parameter_name)
mx_tensor_type = get_ao_tensor(MXTensor, weight.device.type)

quantized_weight = mx_tensor_type.to_mx(
    weight,
    config.weight_dtype,
    block_size=config.block_size,
    kernel_preference=config.kernel_preference,
    act_quant_kwargs=act_quant_kwargs,
    scaling_mode=config.scaling_mode,
)
```

Dispatch should happen at the narrowest stable factory boundary, such as `from_hp`, `to_mx`, or a quantization workflow handler. It should not be repeated inside every operator implementation.

Factories should construct `cls(...)` rather than a hard-coded base class where practical. Structural operations that return the same logical tensor type, such as transpose, view, slice, clone, collectives, and unflatten, must preserve `self.__class__`. This lets inherited generic behavior preserve a device-specific subclass without duplicating the operation in every backend.

### Device-specific subclass responsibilities

A device subclass may override only behavior that differs from the generic implementation, including:

- physical data or scale layout;
- kernel selection or operator arguments;
- hardware capability validation;
- device-specific serialization metadata;
- structural operations when generic `self.__class__` preservation is insufficient.

Shared quantization semantics, public configuration, logical tensor metadata, dequantization behavior, and device-independent operators remain in the generic class.

## MXTensor Example

MXTensor demonstrates why this abstraction is needed. CUDA and XPU share the logical MX format and the public `MXDynamicActivationMXWeightConfig`, but their kernel layout contracts differ.

### Before: device branches in generic code

The implementation in [pytorch/ao#4468](https://github.com/pytorch/ao/pull/4468) enables XPU by extending generic MX code:

- The public configuration gains `swizzle_scales`, with CUDA using `True` and XPU requiring `False`.
- The generic workflow rejects swizzled scales on XPU.
- `_addmm_mx_dispatch` determines `is_xpu` from operands.
- Scale preparation chooses blocked/swizzled CUDA layout or contiguous row-major XPU layout.
- MXFP8 transposes the XPU weight scale to satisfy the XPU `torch._scaled_mm` contract.
- MXFP4 passes `swizzle=None` on XPU and `SWIZZLE_32_4_4` on CUDA.

In simplified form:

```python
is_xpu = a.qdata.is_xpu and b.qdata.is_xpu

if is_xpu:
    a_scale = a.scale.contiguous()
    b_scale = b.scale.contiguous()
else:
    a_scale = to_blocked(a.scale)
    b_scale = to_blocked(b.scale)

swizzle = None if is_xpu else SwizzleType.SWIZZLE_32_4_4
```

This works, but it makes the generic MX implementation responsible for both CUDA and XPU physical contracts.

### After: XPU behavior in an MXTensor subclass

#4800 With device-specific tensor dispatch, the generic factory selects `MXTensorXPU` for XPU inputs. `MXTensorXPU` inherits logical MX behavior and overrides the XPU-specific parts:

```python
class MXTensorXPU(MXTensor):
    @staticmethod
    @torch._dynamo.allow_in_graph
    def to_mx(cls, data_hp, elem_dtype, ...):
        scale, qdata = to_mx(
            data_hp,
            elem_dtype,
            ...,
            is_swizzled_scales=False,
        )
        return cls(qdata, scale, ..., is_swizzled_scales=False)


xpu_implements = MXTensorXPU.implements


@xpu_implements([aten.addmm.default])
def xpu_mx_addmm(func, types, args, kwargs):
    # Use row-major 1x32 scales and the XPU kernel contract.
```

The class name `MXTensorXPU` is an implementation detail. Users continue to request MX quantization through the existing torchao API.

## Backward Compatibility

This proposal is backward compatible by construction:

1. If no backend registers a subclass, `get_ao_tensor(GenericTensor, device)` returns `GenericTensor`. Existing tensor classes do not need to register themselves.
3. Existing CUDA and CPU factory behavior remains unchanged.
4. Public quantization configs and workflow calls do not change.
5. Existing checkpoints containing the generic tensor class continue to load as before.

## Ownership and Maintenance

XPU-specific implementations should live in an `xpu/` subdirectory close to the generic feature. Intel maintainers own these files and their tests through `CODEOWNERS`.

**Comment by vkuzo** [comment]
I believe that the right way to configure numerics and layouts is from user configuration, not automatic branching based on tensor device.  I think the code below is confusing, as automatically changing tensor layouts based on device is unintuitive and complex:

```python
is_xpu = a.qdata.is_xpu and b.qdata.is_xpu

if is_xpu:
    a_scale = a.scale.contiguous()
    b_scale = b.scale.contiguous()
else:
    a_scale = to_blocked(a.scale)
    b_scale = to_blocked(b.scale)

swizzle = None if is_xpu else SwizzleType.SWIZZLE_32_4_4
```

I believe a simpler way to do it instead is this:

```python
# user specifies a spec
WorkflowConfig(..., swizzle=swizzle)

# if needed, provide per-device defaults of workflow config
# something like this can be sugar to set device-specific layout settings
xpu_config = WorkflowConfig.get_default_by_device("xpu")

# modeling code honors the spec specified by the user
# modeling code - this should not branch on numerics/layouts/etc based on tensor device, only on config
if tensor.saved_config.use_swizzle:
    ... swizzle
else:
    ... don't swizzle
```

If there is a need to modify layouts (i.e. load a checkpoint saved for CUDA on XPU), I'd favor explicit code that changes the relevant stored config settings to make it work, instead of dynamically branching at runtime.

**Comment by xiaowangintel** [comment]
> I think the code below is confusing, as automatically changing tensor layouts based on device is unintuitive and complex:

That makes sense to me.  However, the current default behavior in mx_tensor requires a swizzled layout( https://github.com/pytorch/ao/blob/main/torchao/prototype/mx_formats/mx_tensor.py#L789). Even when is_swizzled_scales=False, mx_tensor still changes the tensor layout to swizzled. This works well for CUDA, but it is not friendly for other backends. It is precisely because of the above design, and because we want to preserve the existing CUDA default behavior, that we use the device to determine the tensor layout.

**Comment by liangan1** [comment]
Thanks for your feedback @vkuzo. 
But this RFC is targeting a general idea to maintain the device-specific changes, not just for MXTensor. The current priority is the MXTensor enabling for XPU. Let us firstly focus on the MXTensor.

**Comment by liangan1** [comment]
Thanks for your feedback @vkuzo.
In fact, As for the MXTensor,  Let me try to workout a right path for swizzle support. 

AO is based on the PyTorch scaled_mm. PyTorch define the torch.nn.functional.SwizzleType and scaled_mm API use this type to decide tensor layout. I suppose the right thing to do in the torchAO is based this SwizzleType  to decide the tensor layout. Since the SwizzleType  in the PyTorch is an ENUM type, we should follow it. 

https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/BlasBackend.h#L42
```C++
enum class SwizzleType : std::uint8_t { NO_SWIZZLE = 0, SWIZZLE_32_4_4 = 1 };
```
Then pseudo code  in the AO should as following:

```python
  from torch.nn.functional  import SwizzleType
  # user specifies a spec for CUDA
  WorkflowConfig(..., swizzle_typy=SwizzleType.SWIZZLE_32_4_4)
  # user specifies a spec  for XPU 
  WorkflowConfig(..., swizzle_typy=SwizzleType.NO_SWIZZLE)

  #code in the MXTensor 
   switch(swizzle_type):

   case: SwizzleType.SWIZZLE_32_4_4
   ... swizzle path 

   case:  SwizzleType.NO_SWIZZLE
   ... non swizzle path 
 ```

With this changes, their will no device specific path and it also allow the siwzzle type extension for the future. 
  
But we need to change the current AO code to reache this goal. The current design of MXTensor should be compatible with both swizzle and non-swizzle. But I suppose the current implementation of non-swizzle path should not be the expected behavior. 

In the following code, both swizzle and non-swizzle use to_blokced to decide the layout of weight which means that the swizzle block layout is always used for both configs.  I suppose it should not be expected? whether we can remove the the maybe_dtensor_to_block for non swizzle path?

https://github.com/pytorch/ao/blob/08f8d490ab566553dadb74de1e8430f156da42bc/torchao/prototype/mx_formats/mx_tensor.py#L781-L801

As for the impact, QuantizeTensorToMXKwargs is only used in the inference workflow with  swizzle config and  and non swizzle path is not used by the AO yet. 
https://github.com/search?q=repo%3Apytorch%2Fao+QuantizeTensorToMXKwargs&type=code

**Comment by xiaowangintel** [comment]
Closing this Issue
